src/block_cache/block_chunkcache.c

```diff
@@ -0,0 +1,561 @@
+     *    That prevents the code that removes outdated chunks from freeing the chunk before we do.
+     * 2. Remove the chunk from its chunk's chain, acquiring appropriate locks.
+     * 3. Free the chunk.
+     */
+    __wt_spin_lock(session, &chunkcache->chunkcache_lru_lock);
+    TAILQ_FOREACH_REVERSE(
+      chunk_to_evict, &chunkcache->chunkcache_lru_list, __wt_chunkcache_lru, next_lru_item)
+    {
+        if (chunk_to_evict->valid) {
+            TAILQ_REMOVE(&chunkcache->chunkcache_lru_list, chunk_to_evict, next_lru_item);
+            chunk_to_evict->being_evicted = true;
+            found_eviction_candidate = true;
+            break;
+        }
+    }
+    __wt_spin_unlock(session, &chunkcache->chunkcache_lru_lock);
+
+    if (!found_eviction_candidate)
+        return;
+
+    __wt_verbose(session, WT_VERB_CHUNKCACHE, "evict: %s(%u), offset=%" PRIu64 ", size=%" PRIu64,
+      (char *)&chunk_to_evict->hash_id.objectname, chunk_to_evict->hash_id.objectid,
+      (uint64_t)chunk_to_evict->chunk_offset, (uint64_t)chunk_to_evict->chunk_size);
+
+    __wt_spin_lock(session, WT_BUCKET_LOCK(chunkcache, chunk_to_evict->bucket_id));
+    TAILQ_REMOVE(
+      WT_BUCKET_CHUNKS(chunkcache, chunk_to_evict->bucket_id), chunk_to_evict, next_chunk);
+    __wt_spin_unlock(session, WT_BUCKET_LOCK(chunkcache, chunk_to_evict->bucket_id));
+    __chunkcache_free_chunk(session, chunk_to_evict);
+
+    WT_STAT_CONN_INCR(session, chunk_cache_chunks_evicted);
+}
+
+/*
+ * __chunkcache_eviction_thread --
+ *     Periodically sweep the cache and evict chunks at the end of the LRU list.
+ */
+static WT_THREAD_RET
+__chunkcache_eviction_thread(void *arg)
+{
+    WT_CHUNKCACHE *chunkcache;
+    WT_SESSION_IMPL *session;
+
+    session = (WT_SESSION_IMPL *)arg;
+    chunkcache = &S2C(session)->chunkcache;
+
+    while (!chunkcache->chunkcache_exiting) {
+        /* Try evicting a chunk if we have exceeded capacity. */
+        while (!chunkcache->chunkcache_exiting &&
+          ((chunkcache->bytes_used + chunkcache->chunk_size) >
+            chunkcache->evict_trigger * chunkcache->capacity / 100))
+            __chunkcache_evict_one(session);
+        __wt_sleep(0, 100 * WT_THOUSAND); /* may need tuning */
+    }
+    return (WT_THREAD_RET_VALUE);
+}
+
+/*
+ * __wt_chunkcache_get --
+ *     If the cache has the data at the given size and offset, copy it into the supplied buffer.
+ *     Otherwise, read and cache the chunks containing the requested data.
+ */
+int
+__wt_chunkcache_get(WT_SESSION_IMPL *session, WT_BLOCK *block, uint32_t objectid, wt_off_t offset,
+  uint32_t size, void *dst)
+{
+    WT_CHUNKCACHE *chunkcache;
+    WT_CHUNKCACHE_CHUNK *chunk;
+    WT_CHUNKCACHE_HASHID hash_id;
+    WT_DECL_RET;
+    bool chunk_cached;
+    unsigned int bucket_id, retries;
+    size_t already_read, left_to_read, readable_in_chunk, size_copied;
+
+    chunkcache = &S2C(session)->chunkcache;
+    already_read = 0;
+    left_to_read = size;
```

Comment:
For some reason when I see `left_to_read` I initially say in my head `left to right` and then have to fix my comprehension. What about naming it `remaining` or something like that?